miden-tx/src/tx_benches/utils.rs

```diff
@@ -0,0 +1,267 @@
+    pub fn new(account: AccountStub, adv_provider: A) -> Self {
+        Self {
+            host: TransactionHost::new(account, adv_provider),
+            tx_progress: TransactionProgress::default(),
+        }
+    }
+}
+
+impl<A: AdviceProvider> Host for BenchHost<A> {
+    fn get_advice<S: ProcessState>(
+        &mut self,
+        process: &S,
+        extractor: AdviceExtractor,
+    ) -> Result<HostResponse, ExecutionError> {
+        self.host.get_advice(process, extractor)
+    }
+
+    fn set_advice<S: ProcessState>(
+        &mut self,
+        process: &S,
+        injector: AdviceInjector,
+    ) -> Result<HostResponse, ExecutionError> {
+        self.host.set_advice(process, injector)
+    }
+
+    fn on_trace<S: ProcessState>(
+        &mut self,
+        process: &S,
+        trace_id: u32,
+    ) -> Result<HostResponse, ExecutionError> {
+        let event = TransactionTrace::try_from(trace_id)
+            .map_err(|err| ExecutionError::EventError(err.to_string()))?;
+
+        use TransactionTrace::*;
+        match event {
+            PrologueStart => self.tx_progress.prologue.set_start(process.clk()),
+            PrologueEnd => self.tx_progress.prologue.set_end(process.clk()),
+            NotesProcessingStart => self.tx_progress.notes_processing.set_start(process.clk()),
+            NotesProcessingEnd => self.tx_progress.notes_processing.set_end(process.clk()),
+            NoteConsumingStart => {
+                self.tx_progress.note_consuming.push(CycleInterval::new(process.clk()))
+            },
+            NoteConsumingEnd => {
+                if let Some(interval) = self.tx_progress.note_consuming.last_mut() {
+                    interval.set_end(process.clk())
+                }
+            },
+            TxScriptProcessingStart => {
+                self.tx_progress.tx_script_processing.set_start(process.clk())
+            },
+            TxScriptProcessingEnd => self.tx_progress.tx_script_processing.set_end(process.clk()),
+            EpilogueStart => self.tx_progress.epilogue.set_start(process.clk()),
+            EpilogueEnd => self.tx_progress.epilogue.set_end(process.clk()),
+            ExecutionEnd => {
+                self.tx_progress.print_stages();
+            },
+        }
+
+        Ok(HostResponse::None)
+    }
+
+    fn on_event<S: ProcessState>(
+        &mut self,
+        process: &S,
+        event_id: u32,
+    ) -> Result<HostResponse, ExecutionError> {
+        self.host.on_event(process, event_id)
+    }
+}
+
+// TRANSACTION PROGRESS
+// ================================================================================================
+
+/// Contains the information about the number of cycles for each of the transaction execution
+/// stages.
+#[derive(Default)]
+struct TransactionProgress {
+    prologue: CycleInterval,
+    notes_processing: CycleInterval,
+    note_consuming: Vec<CycleInterval>,
+    tx_script_processing: CycleInterval,
+    epilogue: CycleInterval,
+}
```

Comment:
I would probably expose dedicated methods for setting different properties. For example:
```Rust
impl TransactionProgress {
    pub fn start_prologue(&mut self, cycle: u32) {
        todo!()
    }

    pub fn start_note_processing(&mut self, cycle: u32) {
        todo!()
    }

    pub fn start_note_execution(&mut self, cycle: u32, note_id: NoteId) {
        todo!()
    }

    pub fn start_tx_script_execution(&mut self, cycle: u32) {
        todo!()
    }

    pub fn start_epilogue(&mut self, cycle: u32) {
        todo!()
    }

    pub fn end_transaction(&mut self, cycle: u32) {
        todo!()
    }
}
```

Also, a general question: we seem to be capturing both start and end points for many events - is this really needed? Usually start of one event is the end of the other and something like above should be sufficient (and I'm not even sure `start_prologue()` is needed - isn't it always ~0?).

Reply:
>we seem to be capturing both start and end points for many events - is this really needed?

I can do that (actually, my first implementation was doing so), but in that case it will be more difficult to print the output (the result of notes processing and note execution, specifically): I will get the notes processing interval length (overall length) only at the end of the execution of all notes (it is not a big problem, but IMO the current representation with nested list looks better). But, which is more important, it will be more difficult to detect cases where the ending trace is missing. Now I can handle such cases and print `invalid interval`, but if I will need to calculate the intervals on the fly, I won't print anything since the end of the interval will be never found.
Also I thought that this approach will be more reliable if we will add some other instructions which we don't want to include in the benchmark. 

>and I'm not even sure start_prologue() is needed - isn't it always ~0?

For some reason (which honestly is still not clear for me) start of the prologue is happened at the 11th cycle, so I need it to be able to calculate the prologue's length correctly. Probably it is not 0 because entering the `main` function consumes some cycles?

Reply:
Yep - it's fine to keep more granular start/end events if they simplify things.